lib/spack/spack/database.py

```diff
@@ -885,19 +886,26 @@ def query(
             end_date (datetime, optional): filters the query discarding
                 specs that have been installed after ``end_date``.
 
+            hashes (container): list or set of hashes that we can use to
+                restrict the search
+
         Returns:
             list of specs that match the query
+
         """
         # TODO: Specs are a lot like queries.  Should there be a
         # TODO: wildcard spec object, and should specs have attributes
         # TODO: like installed and known that can be queried?  Or are
         # TODO: these really special cases that only belong here?
+
+        # TODO: handling of hashes restriction is not particularly elegant.
```

Comment:
@tgamblin I still think `Database` should not be modified in this PR. I expanded on that at https://github.com/spack/spack/pull/9612#issuecomment-434015313

Reply:
@scheibelp: I didn't see a counterargument to the performance point I made.  The reason I think this should be in `query` is because it really is just another constraint (subset of specs) and `query` is a general query function.

If you keep this outside `query`, then every `query` does as many spec compares as there are installed packages.  If you push the hash comparison *into* `query()`, then we can quickly eliminate specs that aren't in the set of hashes, and only do spec comparisons on the things *in* the subset of hashes.

I could see having two query functions *if* they were separable and you could put the fast one first, but right now that's not possible without a bigger refactor of `Database`.  Thoughts?

Reply:
I'd rather this were refactored into a 2-3 functions (in a separate PR) and that it stayed as simple as possible for now (even at the cost of `find` being a bit slower than it could be in an environment):

* A function with the bulk of this logic that doesn't use `self._data` and just looks through a set of specs that it is given
* A function that runs the first function with a set of specs

Then you can call the second function with `self._data` or "the set of specs in an environment"

IMO this is the sort of speedup that could be added later: I'd prefer to reduce complexity added by this PR since it adds quite a lot already.